src/Storages/StorageMergeTreeIndex.cpp

```diff
@@ -0,0 +1,307 @@
+            {
+                result_columns[pos] = fillMarks(part, marks_loader, *column_type, first);
+            }
+            else
+            {
+                throw Exception(ErrorCodes::NO_SUCH_COLUMN_IN_TABLE, "No such column {}", column_name);
+            }
+        }
+
+        ++part_index;
+        return Chunk(std::move(result_columns), num_rows);
+    }
+
+private:
+    std::shared_ptr<MergeTreeMarksLoader> createMarksLoader(const MergeTreeDataPartPtr & part, const String & prefix_name, size_t num_columns)
+    {
+        auto info_for_read = std::make_shared<LoadedMergeTreeDataPartInfoForReader>(part, std::make_shared<AlterConversions>());
+        auto local_context = getContext();
+
+        return std::make_shared<MergeTreeMarksLoader>(
+            info_for_read,
+            local_context->getMarkCache().get(),
+            info_for_read->getIndexGranularityInfo().getMarksFilePath(prefix_name),
+            info_for_read->getMarksCount(),
+            info_for_read->getIndexGranularityInfo(),
+            /*save_marks_in_cache=*/ false,
+            local_context->getReadSettings(),
+            /*load_marks_threadpool=*/ nullptr,
+            num_columns);
+    }
+
+    ColumnPtr fillMarks(
+        MergeTreeDataPartPtr part,
+        std::shared_ptr<MergeTreeMarksLoader> marks_loader,
+        const IDataType & data_type,
+        const String & column_name)
+    {
+        size_t col_idx = 0;
+        bool has_marks_in_part = false;
+        size_t num_rows = part->index_granularity.getMarksCount();
+
+        if (isWidePart(part))
+        {
+            if (auto stream_name = part->getStreamNameOrHash(column_name, part->checksums))
+            {
+                col_idx = 0;
+                has_marks_in_part = true;
+                marks_loader = createMarksLoader(part, *stream_name, /*num_columns=*/ 1);
+            }
+        }
+        else if (isCompactPart(part))
+        {
+            if (auto col_idx_opt = part->getColumnPosition(column_name))
+            {
+                col_idx = *col_idx_opt;
+                has_marks_in_part = true;
+            }
+        }
+        else
+        {
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Parts with type {} are not supported", part->getTypeName());
+        }
+
+        if (!has_marks_in_part)
+        {
+            auto column = data_type.createColumnConstWithDefaultValue(num_rows);
+            return column->convertToFullColumnIfConst();
+        }
+
+        auto compressed = ColumnUInt64::create(num_rows);
+        auto uncompressed = ColumnUInt64::create(num_rows);
+
+        auto & compressed_data = compressed->getData();
+        auto & uncompressed_data = uncompressed->getData();
+
+        for (size_t i = 0; i < num_rows; ++i)
+        {
+            auto mark = marks_loader->getMark(i, col_idx);
+
+            compressed_data[i] = mark.offset_in_compressed_file;
+            uncompressed_data[i] = mark.offset_in_decompressed_block;
```

Comment:
And row count for adaptive granularity?

Reply:
It's already added (column `rows_in_granule`)